### ISSUE TITLE
DNS query crash when connecting to ES Cloud

### ISSUE THREAD

**Issue by alexeyzimarev** [pyeventsourcing/esdbclient#17]
I am not sure if it's related to ES Cloud only (it shouldn't) but I have a report of an exception:

> Traceback (most recent call last):
File "/home/ubuntu/.local/lib/python3.10/site-packages/esdbclient/client.py", line 204, in _connect_to_preferred_node
answers = dns.resolver.resolve(cluster_fqdn, "A")
File "/home/ubuntu/.local/lib/python3.10/site-packages/dns/resolver.py", line 1565, in resolve
return get_default_resolver().resolve(
File "/home/ubuntu/.local/lib/python3.10/site-packages/dns/resolver.py", line 1307, in resolve
(request, answer) = resolution.next_request()
File "/home/ubuntu/.local/lib/python3.10/site-packages/dns/resolver.py", line 749, in next_request
raise NXDOMAIN(qnames=self.qnames_to_try, responses=self.nxdomain_responses)
dns.resolver.NXDOMAIN: The DNS query name does not exist: clt07f5o0aesbufe1n6g.mesdb.eventstore.cloud:2113.

I checked the line 204 and it calls `answers = dns.resolver.resolve(cluster_fqdn, "A")`. From the error message I understand that `cluster_fqdn` also contains the port, which is apparently wrong.

The code that calls the client looks totally fine, so I am not sure what went wrong:

```python
client = EventStoreDBClient(
    uri="esdb+discover://admin:something@<cloud-cluster-id>.mesdb.eventstore.cloud:2113"
)
```

We tried to use `esdb://<cloud-cluster-id>-0.mesdb.eventstore.cloud:2113` as I had a guess that `esdb+discover` might be an issue, but it crashed with the same error. It tried to resolve `<cloud-cluster-id>-0.mesdb.eventstore.cloud:2113`, again, including the port in DNS query.

**Comment by alexeyzimarev** [comment]
I think what I wrote is wrong, it should work with `esdb://`. Basically, as I can see here:

```python
            cluster_fqdn = self.connection_spec.targets[0]
            try:
                answers = dns.resolver.resolve(cluster_fqdn, "A")
            except dns.exception.DNSException as e:
                raise DNSError() from e
            gossip_seed: Sequence[str] = [f"{s.address}:2112" for s in answers]
```

there are several issues.
- The `targets` list contain the host name _and_ the port, so calling `resolve` with host:port crashes
- Using DNS for getting the gossip response from a cluster DNS is not right. Instead, you should use the client gossip gRPC call using the cluster DNS as an endpoint, and use addresses returned instead of creating a gossip seed using IP addresses and hard-coded `2112` port. It's not guaranteed that IP addresses will work (certs CN is the hostname), and client advertise as address doesn't necessarily matches the DNS response.

**Comment by alexeyzimarev** [comment]
Basically, there should be no calls to resolve DNS names to IPs in the client. Instead, there need to be a call to this service: https://github.com/pyeventsourcing/esdbclient/blob/54e768e26ed7915b7e7d7b7648c0c1fbfb5b9984/protos/esdbclient/protos/Grpc/gossip.proto#L7-L9
The result returned from there should be used as-is. Of course, the client would need to choose the right node to connect to, using the node preference from the connection string, and ensuring the node isn't dead. The NodeJS client Is a good example: https://github.com/EventStore/EventStore-Client-NodeJS/blob/91fb2aac898ea2c6dce072f94393466390ccf89c/src/Client/discovery.ts

**Comment by johnbywater** [comment]
Thanks for raising this @alexeyzimarev.

I'm not saying you are wrong and I'm not asserting the client is right. But what you describe here is a totally different approach to what this client has been coded to do. I have been carefully following what Yves and others told me to do with the "esdb+discover" scheme. And this has been reviewed by several people.

It's been [clearly documented](https://github.com/pyeventsourcing/esdbclient#two-schemes): that (1) The "esdb+discover" scheme has a "cluster domain name" and does not have a port number; that (2) that this "cluster domain name" is used to get A records from a DNS server, and the hosts mentioned in these A recorded used as "grpc targets" with a hard coded port number of 2113; that (3) "cluster info" is obtained from the "grpc targets" and then the client may reconnect according to the node preference (if the node preference is already satisfied then it won't reconnect).

I am of course more than happy to review and revise. I don't fully understand what you have said above, the issues of connecting to a cluster, connecting to the cloud service, and using the two schemes are a bit mixed together, and altogether I'm a bit confused about what you are saying. Let me read it all through again and see if I can make more sense of it?

**Comment by alexeyzimarev** [comment]
I am sorry to say that those guidelines that you received aren't correct _today_.

First, the connection string used in the sample I provided is correct, and it works in any other client. There's no requirement that when using cluster DNS the port _must be_ 2113, and it cannot be provided in the URI. So, that connection string should work, it's completely legit.

Second, the procedure to discover the node for the client to connect to should work like this (it's _relatively_ new):
- Use the hostname and port part of the URI of `esd+discover` schema
- Create an instance of the `GossipService` generated from `gossip.proto`. Use the URI from the previous step.
- Make a call to that service's `Read` function and get the response back
- Use the list of retuned nodes as the list of nodes (gossip seed if you like)
- In that list, every node has the `http_end_point` property (among other helpful things like node role and liveness), which must be used to populate the "gossip seed". That property includes hostname and port unconditionally, so there's no guesswork.

The rest of the discovery would work as far as I was able to understand.

**Comment by johnbywater** [comment]
@alexeyzimarev I have made the superficial change of allowing the esdb+discover scheme to have an optional port-number after the cluster-domainname, and this is not used when making the DNS call, so that the DNS call isn't broken by including a port number in the URI, and then also this port number is used when constructing grpc targets from the A records, and if the URI does not mention a port number the default port number of 2113 is used. I noticed that the code actually had 2112 as the default, for some reason. It was documented as 2113 but the code was actually using 2112, which was a mistake. These changes are in v1.0.13.

**Comment by johnbywater** [comment]
I know that isn't what you asked for, I just thought I would throw that in with the other changes in v1.0.13 because it was an easy improvement.

Regarding the intentions for the esdb+discover scheme, I really don't understand the difference between esdb+discover and esdb. The esdb scheme already uses the list of grpc targets to get the gossip and then select a node according to the cluster info and the node preference. The redundancy in having a list of targets seems to increase the changes of reaching the cluster. If the esdb+discover scheme only mentions one node, that node might not be available. But DNS servers are usually quite reliable, probably more reliable than an EventStoreDB server, since they have only one job, and you can configure alternatives etc. And decoupling the cluster node names from the name of the cluster via DNS seems like a good idea to me. And also if esdb+discover is using "cluster DNS" that isn't actually DNS as we know it, why use the term DNS at all? And if "esdb" is to use cluster info from the gossip service then I don't see the difference with the semantics you described for "esdb+discover". And if "esdb" is to not use cluster info from the gossip service then it just opens possibilities for expressing a node preference in the URI but specifying a node in the URI that doesn't satisfy this node preference. Maybe we could have a meeting about this? Or discuss it on the Discord channel, or maybe there's already a long thread about this? Who is responsible for designing this? Is what I have coded a correct understanding of an older design? If it has changed, why did it change? I'm quite confused.

**Comment by alexeyzimarev** [comment]
I have to add that, indeed, the current implementation is based on _correct_ requirements. What is implemented in the Python client is how our clients worked about two years ago.

Regarding the difference between `esdb` and `esdb+discover`. When `esdb` is used, there's no need to do any discovery. You have one host, and you don't need to call the gossip endpoint to get the list of nodes. You don't need to check if the node is alive because you have no option to connect to another node if the original node is dead. Node preference doesn't apply. So, basically, there's no "discovery" process going on there.

For example, one can unconditionally connect to a read-only replica if they know its hostname. They won't use `esdb+discover` with node preference set to RoR, but instead just connect directly. If they'd use `esdb+discover', the client would _try_ connecting to an RoR but if it's not available, the client would connect to some other node, and that might not be what the user wants.

**Comment by johnbywater** [comment]
This was fixed in b30f99de9ad1805a65d9ce4eb9a7d1da46531cc2 released as version 1.0.18.
